lib/SDIR/Translate/translateToSDFG.cpp

```diff
@@ -0,0 +1,1597 @@
+      if (translateToSDFG(*alloc, jemit).failed())
+        return failure();
+
+    if (AllocTransientStreamOp alloc = dyn_cast<AllocTransientStreamOp>(oper))
+      if (translateToSDFG(*alloc, jemit).failed())
+        return failure();
+  }
+
+  for (Operation &oper : op.body().getOps())
+    if (StateNode state = dyn_cast<StateNode>(oper)) {
+      for (AllocOp allocOper : state.getAllocs())
+        if (translateToSDFG(*allocOper, jemit).failed())
+          return failure();
+
+      for (AllocTransientOp allocOper : state.getTransientAllocs())
+        if (translateToSDFG(*allocOper, jemit).failed())
+          return failure();
+
+      for (AllocStreamOp allocOper : state.getStreamAllocs())
+        if (translateToSDFG(*allocOper, jemit).failed())
+          return failure();
+
+      for (AllocTransientStreamOp allocOper : state.getTransientStreamAllocs())
+        if (translateToSDFG(*allocOper, jemit).failed())
+          return failure();
+    }
+  jemit.endObject(); // _arrays
+
+  jemit.startNamedObject("symbols");
+  for (Operation &oper : op.body().getOps()) {
+    if (AllocSymbolOp alloc = dyn_cast<AllocSymbolOp>(oper))
+      if (translateToSDFG(*alloc, jemit).failed())
+        return failure();
+  }
+
+  jemit.endObject(); // symbols
+
+  if (!containsAttr(*op, "instrument"))
+    jemit.printKVPair("instrument", "No_Instrumentation");
+
+  jemit.startNamedObject("init_code");
+  jemit.startNamedObject("frame");
+  jemit.printKVPair("string_data", "");
+  jemit.printKVPair("language", "CPP");
+  jemit.endObject(); // frame
+  jemit.endObject(); // init_code
+
+  jemit.startNamedObject("exit_code");
+  jemit.startNamedObject("frame");
+  jemit.printKVPair("string_data", "");
+  jemit.printKVPair("language", "CPP");
+  jemit.endObject(); // frame
+  jemit.endObject(); // exit_code
+
+  jemit.startNamedObject("global_code");
+  jemit.startNamedObject("frame");
+  jemit.printKVPair("string_data", "");
+  jemit.printKVPair("language", "CPP");
+  jemit.endObject(); // frame
+  jemit.endObject(); // global_code
+
+  jemit.printKVPair("name", op.sym_name());
+  jemit.endObject(); // attributes
+
+  jemit.startNamedList("nodes");
+
+  unsigned stateID = 0;
+
+  for (Operation &oper : op.body().getOps())
+    if (StateNode state = dyn_cast<StateNode>(oper)) {
+      state.setID(stateID);
+      if (translateStateToSDFG(state, jemit).failed())
+        return failure();
+      stateID++;
+    }
+
+  jemit.endList(); // nodes
+
+  jemit.startNamedList("edges");
+
+  for (Operation &oper : op.body().getOps())
+    if (EdgeOp edge = dyn_cast<EdgeOp>(oper))
+      if (translateToSDFG(*edge, jemit).failed())
+        return failure();
+
+  jemit.endList(); // edges
+
+  StateNode entryState = op.getStateBySymRef(op.entry());
+  jemit.printKVPair("start_state", entryState.ID(), /*stringify=*/false);
+
+  return success();
+}
+
+LogicalResult translateSDFGToSDFG(SDFGNode &op, JsonEmitter &jemit) {
```

Comment:
Name is a bit misleading

Reply:
Suggestions?
`translateSDFGNodeToSDFG` ?

Reply:
:/

Reply:
My best suggestion, name-wise, is to try to use a polymorphic `.translateToSDFG()`